### PR TITLE
pypi: stop excluding setuptools

### PR DESCRIPTION
**Pull request by branchvincent**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

---

Python 3.12 no longer bundles setuptools (see https://docs.python.org/3/whatsnew/3.12.html#ensurepip), so `brew update-python-resources` shouldn't exclude it anymore for those versions

See also https://github.com/orgs/Homebrew/discussions/5028, https://github.com/Homebrew/homebrew-core/pull/159180

### CODE REVIEW

File: Library/Homebrew/utils/pypi.rb [hunk 1]
```diff
@@ -249,9 +249,16 @@ def self.update_python_resources!(formula, version: nil, package_name: nil, extr
       end
     end
 
+    python_deps = formula.deps
+                         .map(&:to_formula)
+                         .select { |f| f.name.start_with?("python@") }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:254]:
```suggestion
                         .select { |f| f.name.start_with?("python@") }
                         .map(&:to_formula)
```

as `to_formula` is somewhat expensive and can fail

File: Library/Homebrew/utils/pypi.rb [hunk 2]
```diff
@@ -249,9 +249,16 @@ def self.update_python_resources!(formula, version: nil, package_name: nil, extr
...
+                         .sort_by(&:version)
+                         .reverse
+
     extra_packages = (extra_packages || []).map { |p| Package.new p }
     exclude_packages = (exclude_packages || []).map { |p| Package.new p }
-    exclude_packages += %w[argparse pip setuptools wsgiref].map { |p| Package.new p }
+    exclude_packages += %w[argparse pip wsgiref].map { |p| Package.new p }
+    exclude_packages.append(Package.new("setuptools")) if python_deps.first&.version&.< Version.new("3.12")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:261]: How do we know `.first` is the right one here? Might be worth doing something like this too:
```suggestion
    exclude_packages.append(Package.new("setuptools")) if python_deps.first && python_deps.first.version < Version.new("3.12")
```

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks @branchvincent, a few comments (not all your fault too!)

File: Library/Homebrew/utils/pypi.rb (hunk 1 above)
Comment (branchvincent) [reply Library/Homebrew/utils/pypi.rb:254]: done, had to switch to a regex to handle `uses_from_macos "python"`

File: Library/Homebrew/utils/pypi.rb (hunk 2 above)
Comment (branchvincent) [reply Library/Homebrew/utils/pypi.rb]: `python_deps` is reverse sorted by `version`, so first should be the max which is what we want

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks again @branchvincent!

File: Library/Homebrew/utils/pypi.rb [hunk 3]
```diff
@@ -249,9 +249,18 @@ def self.update_python_resources!(formula, version: nil, package_name: nil, extr
...
+                         .select { |d| d.name.match?(/^python(@.+)?$/) }
+                         .map(&:to_formula)
+                         .sort_by(&:version)
+                         .reverse
+
     extra_packages = (extra_packages || []).map { |p| Package.new p }
     exclude_packages = (exclude_packages || []).map { |p| Package.new p }
-    exclude_packages += %w[argparse pip setuptools wsgiref].map { |p| Package.new p }
+    exclude_packages += %w[argparse pip wsgiref].map { |p| Package.new p }
+    if python_deps.first && python_deps.first.version < Version.new("3.12")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/pypi.rb:261]:
```suggestion
    if (newest_python = python_deps.first) && newest_python.version < Version.new("3.12")
```

**Pull request merged**